---
title: "bsa_strain_genotype_check"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{bsa_strain_genotype_check}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r setup}
library(BSA)
library(here)
library(tidyverse)
library(foreach)
library(GenomicRanges)
library(BSgenome.CneoformansVarGrubiiKN99.NCBI.ASM221672v1)
library(SeqArray)
library(SeqVarTools)
library(Rsamtools)

kn99_genome = BSgenome.CneoformansVarGrubiiKN99.NCBI.ASM221672v1
```

```{r}
input_vcf_tables = list.files(
  "/mnt/scratch/genotype_check/run_21_wgs_result_markers/variants/filtered/tables",
  full.names = TRUE)
names(input_vcf_tables) = str_remove(basename(input_vcf_tables), "_freebayes_filtered.tsv")
input_vcf_tables = as.list(input_vcf_tables)

bsa_vcf = "/mnt/scratch/genotype_check/run_21_wgs_results/variants/filtered/1_freebayes_filtered.vcf.gz"

bsa_gds = here("data/parent_strains.gds")

meta_df = read_csv('/home/oguzkhan/Downloads/Doering_lab_WGS - 2024_0208 WGS(1).csv') %>%
  janitor::clean_names()

coordinate_meta_df = meta_df %>%
  mutate(marker1 = ifelse(marker %in% c('G418', 'NAT'), coordinates_if_applicable, '' )) %>%
  mutate(marker1 = ifelse(marker1=='N.A.', '', marker1)) %>%
  separate(marker1, c('marker1_start', 'marker1_end'), sep = ',', remove = FALSE) %>%
  select(-marker1) %>%
  replace_na(list(marker1_start='', marker1_end='')) %>%
  mutate(marker2 = ifelse(marker %in% c('G418 & NAT'), coordinates_if_applicable, '')) %>%
  mutate(marker2 = ifelse(marker2 == 'N.A.', '', marker2)) %>%
  # Assuming marker2 contains the string, now split it into separate columns
  separate(marker2, into = c("marker1_start_tmp", "marker1_end_tmp", "marker2_start", "marker2_end", "region"), sep = ",", extra = "merge", fill = "right") %>%
  mutate(marker1_start = ifelse(marker1_start=='', marker1_start_tmp, marker1_start),
         marker1_end = ifelse(marker1_end=='', marker1_end_tmp, marker1_end)) %>%
  select(-c(marker1_start_tmp,marker1_end_tmp)) %>%
  select(tube_number,marker,marker1_start,marker1_end,marker2_start,marker2_end,region) %>%
  # trim whitespace on all columns
  mutate_all(str_trim)
  
```

```{r}
parse_cigar_for_coordinate <- function(positions, cigars, coordinate) {
  results <- lapply(seq_along(cigars), function(i) {
    pos <- positions[i]
    cigar <- cigars[i]
    
    # Split CIGAR string into operations and lengths
    ops <- gregexpr("[A-Z]", cigar)[[1]]
    lengths <- as.integer(unlist(regmatches(cigar, gregexpr("[0-9]+", cigar))))
    operations <- regmatches(cigar, gregexpr("[A-Z]", cigar))[[1]]
    
    aln_left <- 0
    aln_right <- 0
    reached_coordinate <- FALSE
    
    for (j in seq_along(operations)) {
      op <- operations[j]
      len <- lengths[j]
      
      # For alignment length calculation, consider match, deletion, and mismatch
      if(op %in% c("M", "D", "N", "=", "X")) {
        if (pos + len <= coordinate) {
          aln_left <- aln_left + len
        } else if (pos > coordinate) {
          aln_right <- aln_right + len
        } else {
          aln_left <- aln_left + (coordinate - pos)
          aln_right <- aln_right + (pos + len - coordinate - 1)
          reached_coordinate <- TRUE
          break
        }
        pos <- pos + len
      }
      
      if (reached_coordinate) break
    }
    
    return(c(length_aln_left = aln_left, length_aln_right = aln_right))
  })
  
  return(bind_rows(results))
}

# Define the function
extract_reads_by_alignment_score <- function(sample, bam_file, region, min_alignment_score) {
  # Parse the region
  chrom <- gsub(":.*", "", region)
  pos <- as.numeric(gsub(".*:", "", region))
  
  # Ensure only valid fields are requested
  valid_fields <- scanBamWhat()
  requested_fields <- c("qname", "pos", "cigar", "qual")
  what_fields <- requested_fields[requested_fields %in% valid_fields]
  
  # Define the scan parameters, ensuring only valid fields are included
  param <- ScanBamParam(
    which = GRanges(chrom, IRanges(start = pos, end = pos)),
    what = what_fields,
    tag = c("AS", "NM")
  )
  
  # Read the BAM file
  aln <- scanBam(bam_file, param = param)[[1]]
  
  match_length_left_right <- parse_cigar_for_coordinate(aln$pos, aln$cigar, pos)
  
  tibble(sample=sample,
         chr=chrom,
         pos=pos,
         qname = aln$qname,
         cigar=aln$cigar,
         align_qual=aln$tag$AS,
         edit_dist=aln$tag$NM) %>%
    bind_cols(match_length_left_right)
}

# Example usage
# bam_file <- "/mnt/scratch/genotype_check/run_21_wgs_result_markers/alignment/TH0003_sorted_markdups_tagged.bam"
# region <- "TH_C8_NAT_IR6:301"
# min_alignment_score <- 0  # Example threshold
# reads_df <- extract_reads_by_alignment_score('sample1',bam_file, region, min_alignment_score)

bam_list=list.files("/mnt/scratch/genotype_check/run_21_wgs_result_markers/alignment",
                    pattern="*bam$", full.names=TRUE)
bam_df = tibble(bam=bam_list, tube_number=str_remove(basename(bam_list),'_sorted_markdups_tagged.bam'))
marker_check_df = coordinate_meta_df %>%
  left_join(bam_df) %>%
  select(tube_number, bam, marker1_start, marker1_end, marker2_start, marker2_end) %>%
  pivot_longer(-c(tube_number,bam)) %>%
  filter(complete.cases(.)) %>%
  filter(value != '') %>%
  select(-name) %>%
  dplyr::rename(coordinate=value)

get_reads_over_region_per_sample <- function(tube_number, bam, coordinate, ...) {
  message(sprintf('working on %s', tube_number))
  
  extract_reads_by_alignment_score(tube_number, bam, coordinate, 0)
}

res = pmap(marker_check_df, get_reads_over_region_per_sample)

res_out = res %>% bind_rows() %>%
  group_by(sample)

res_out_names = group_keys(res_out)$sample

res_out_list = res_out %>%
  group_split()

# map2(res_out_list, res_out_names,
#      ~write_csv(.x,
#                 file.path("/mnt/scratch/genotype_check/run_21_wgs_results/marker_check",
#                           paste0(.y,'.csv'))))

res_out %>%
  group_by(sample,chr,pos) %>%
  tally() %>%
  ungroup() %>%
  group_by(sample) %>%
  filter(n() %in% c(1,3))

```


```{r}

read_in_vcf_tables = function(path){
  read_tsv(path) %>% 
    dplyr::rename(depth=DP) %>%
    dplyr::rename_with(.fn = ~ gsub("^(.*?\\.)", "", .x), cols=everything()) %>%
    select(-c(DP, AD, GT)) %>%
    dplyr::rename(alt_depth = AO,
                  ref_depth = RO) %>%
    dplyr::rename(CHR=CHROM)
}

input_vcf_tables_df = map(input_vcf_tables, read_in_vcf_tables) %>%
  bind_rows(.id = "sample")
```


```{r}
tmp_dir = tempdir()
bsa_df = vcf_to_qtlseqr_table(bsa_vcf, tmp_dir,
                              parent_ref_sample = "KN99a",
                              parent_alt_sample="C8",
                              parent_filter=FALSE,
                              overwrite = TRUE)

parent_df = bsa_df %>%
  filter(sample %in% c('KN99a', 'C8')) %>%
  mutate(sample = case_when(sample == 'KN99a' ~ 'KN99a',
                            sample == 'C8' ~ 'C8')) %>%
  select(CHR,POS, REF, ALT1, RealDepth, Alternative1, QUAL, sample,
         genotype, Alt1_percentage, Reference, Ref_percentage)

kn99a_df = parent_df %>%
  filter(sample == 'KN99a') %>%
  pivot_wider(names_from = sample,
              values_from = c(RealDepth, Alternative1, 
                              Reference, QUAL, genotype,
                              Alt1_percentage, Ref_percentage)) %>%
  rename_with(.fn = ~ gsub("_KN99a$", "", .x), .cols = everything()) %>%
  dplyr::rename(parent_genotype=genotype) %>%
  mutate(parent_genotype_strict = ifelse(Alt1_percentage>=.9, 1, 0)) %>%
  mutate(confidently_reference =
           ifelse((parent_genotype==0 & parent_genotype_strict==0),
                  TRUE, FALSE)) %>%
  filter(RealDepth >= 10) %>%
  arrange(desc(Alt1_percentage)) %>%
  mutate(kn99a=TRUE)
  

c8_df = parent_df %>%
  filter(sample == 'C8') %>%
  pivot_wider(names_from = sample,
              values_from = c(RealDepth, Alternative1, 
                              Reference, QUAL, genotype,
                              Alt1_percentage, Ref_percentage)) %>%
  rename_with(.fn = ~ gsub("_C8$", "", .x), .cols = everything()) %>%
  dplyr::rename(parent_genotype=genotype) %>%
  mutate(parent_genotype_strict = ifelse(Alt1_percentage>=.9, 1, 0)) %>%
  mutate(confidently_reference =
           ifelse((parent_genotype==0 & parent_genotype_strict==0),
                  TRUE, FALSE)) %>%
  filter(RealDepth >= 10) %>%
  arrange(desc(Alt1_percentage)) %>%
  mutate(c8 = TRUE)

parent_labeling_df = kn99a_df %>%
  filter(parent_genotype==0) %>%
  mutate(parent = 'KN99a') %>%
  select(-kn99a) %>%
  bind_rows(c8_df %>%
              filter(parent_genotype==1) %>%
              mutate(parent = 'c8') %>%
              select(-c8)) %>%
  select(CHR,POS,ALT1,parent)

```

```{r}
# UNTESTED
# 
# Function to identify the longest run allowing a small number of variants within
find_longest_run_with_tolerance <- function(gt_data, tolerance = 2) {
  runs <- rle(gt_data)
  values <- runs$values
  lengths <- runs$lengths
  
  # Initialize variables to track the longest run allowing tolerance
  max_run_length <- 0
  max_run_start <- 1
  max_run_end <- 1
  
  # Initialize variables for the current run
  current_run_length <- 0
  current_run_start <- 1
  variants_in_run <- 0
  
  for (i in seq_along(lengths)) {
    # If the current segment is reference (1) or a small number of variants within tolerance
    if (values[i] == 1 || (values[i] == 0 && variants_in_run + lengths[i] <= tolerance)) {
      current_run_length <- current_run_length + lengths[i]
      if (values[i] == 0) {
        variants_in_run <- variants_in_run + lengths[i]  # Update count of variants within this run
      }
    } else {
      # End current run if the segment is a variant exceeding the tolerance
      if (current_run_length > max_run_length) {
        max_run_length <- current_run_length
        max_run_end <- sum(lengths[1:i])  # Calculate the end position of the current longest run
        max_run_start <- max_run_end - max_run_length + 1  # Calculate the start position
      }
      # Reset for the next run
      current_run_length <- lengths[i] * (values[i] == 1)
      variants_in_run <- lengths[i] * (values[i] == 0)
    }
  }
  
  # Check if the last run is the longest
  if (current_run_length > max_run_length) {
    max_run_length <- current_run_length
    max_run_end <- sum(lengths)  # The run ends at the last position
    max_run_start <- max_run_end - max_run_length + 1
  }
  
  # Return start, end, and length of the longest run
  list(start = max_run_start, end = max_run_end, length = max_run_length)
}

# Usage with your data
gt_data <- as.integer(assays(region_subset)$GT != "0")  # Assuming '0' is reference; converting to binary 0s and 1s
tolerance = 2  # Allowing up to 2 '0's (variants) in a run
longest_run_info <- find_longest_run_with_tolerance(gt_data, tolerance)

# To apply this to the region_subset and extract the corresponding genomic coordinates:
genomic_start <- start(region_subset)[longest_run_info$start]
genomic_end <- end(region_subset)[longest_run_info$end]

```


```{r}
gatk_table_path = "/mnt/scratch/genotype_check/run_21_wgs_results/output.table"
tmp_meta = tempfile(fileext='.csv')
coordinate_meta_df %>% dplyr::rename(sample=tube_number) %>%
  write_csv(tmp_meta)
col_data_path = tmp_meta
bsae = BSA::makeBSAExperimentFromGatkTable(
  gatk_table_path,
  col_data_path,
  drop_samples = c('C8', 'KN99a')
)

SummarizedExperiment::rowRanges(bsae)$c8 = SummarizedExperiment::rowRanges(bsae) %>% 
  as_tibble() %>% 
  left_join(c8_df %>% 
              filter(parent_genotype==1) %>%
              select(CHR,POS, ALT1, c8),
            by = c('seqnames'='CHR', 'start'='POS', 'alt'='ALT1')) %>% 
  tidyr::replace_na(list(c8=FALSE)) %>%
  pull(c8)

# get percent of strain that is c8
percent_c8 = colSums(assays(bsae[bsae@rowRanges$c8,])$GT) / nrow(bsae[bsae@rowRanges$c8,])


summarize_region_origin = function(sample){

  col_mask = bsae@colData$sample==sample
  region_str = bsae[,col_mask]@colData$region
  
  # The input string
  regex_pattern <- "^(.*):([0-9]+)-([0-9]+)$"
  # Using stringr::str_match to extract the components
  matches <- str_match(region_str, regex_pattern)

  # The first row contains the full match, and subsequent rows contain the captured groups
  chr <- matches[1, 2]  # The chromosome
  start <- as.integer(matches[1, 3])  # The start position
  end <- as.integer(matches[1, 4])    # The end position
  
  row_mask=(bsae@rowRanges@seqnames==chr
            & bsae@rowRanges@ranges@start>=start
            & bsae@rowRanges@ranges@start<=end)
  
  region_subset = bsae[row_mask,col_mask]
  
  # find the number of positions that are not reference
  
  region_percent_c8 = sum(assays(region_subset)$GT)/nrow(region_subset)
  # find the most significant run of non-reference positions. Return
  # the start and end coordinates of that run
  # Compute the lengths and values of runs of equal values in a vector
  # Assuming you have a binary version of GT (1 for non-reference, 0 for reference)
  # Creating a tibble for plotting and fitting
  data_tibble <- tibble(
    x = 1:nrow(assays(region_subset)$GT),
    y = assays(region_subset)$GT[,1]
  )
  
  # Fit logistic regression
  logit_model <- glm(y ~ x, data = data_tibble, family = binomial)
  
  # Predict probabilities
  data_tibble <- data_tibble %>%
      mutate(predicted_probability = predict(logit_model, type = "response"))
  
  # Find the breakpoint, defined as where the predicted probability is closest to 50%
  breakpoint <- data_tibble %>%
      filter(abs(predicted_probability - 0.5) == min(abs(predicted_probability - 0.5))) %>%
      dplyr::slice(1) %>%
      dplyr::pull(x)

  ggplot(data_tibble, aes(x = x, y = y)) +
  geom_point() +
  geom_line(aes(y = predicted_probability), color = "blue") +
  geom_vline(xintercept = breakpoint, color = "red", linetype = "dashed")
  
  # Return the start and end coordinates of that run
  tibble(
    region_percent_c8 = region_percent_c8,
    region_breakpoint_coord = paste0(chr,':',start+breakpoint),
    length_from_left = breakpoint,
    width = (end-start)+1
  )
  
}

region_samples = bsae@colData %>% as_tibble() %>%
  filter(complete.cases(.)) %>%
  pull(sample)
names(region_samples) = region_samples

region_summaries = map(region_samples, summarize_region_origin) %>%
  bind_rows(.id = "sample")

input_vcf_tables_df_list = input_vcf_tables_df %>%
  filter(!startsWith(sample, "DGK")) %>%
  group_by(sample) %>%
  droplevels()

wgs_var_summaries = map(group_split(input_vcf_tables_df_list), summarize_genomic_origin)
names(wgs_var_summaries) = group_keys(input_vcf_tables_df_list)$sample

wgs_var_summaries_df = wgs_var_summaries %>%
  bind_rows(.id = "sample")

unusual_samples = wgs_var_summaries_df %>%
  filter(!var_origin %in% c('denovo', 'both'),
         genome_wide_percent < 20) %>%
  pull(sample)

# map2(wgs_var_summaries, group_keys(input_vcf_tables_df_list)$sample,
#      ~write_csv(.x, file.path('/mnt/scratch/genotype_check/run_21_wgs_results/wgs_variant_summary',
#                           paste0(.y, '.csv'))))

wgs_var_summaries_df %>%
  filter(sample %in% unusual_samples) %>%
  left_join(coordinate_meta_df %>% select(tube_number, marker), by=c('sample'='tube_number')) %>%
  View()

# regions samples
region_df = coordinate_meta_df %>%
  filter(!is.na(region)) %>%
  select(tube_number, region) %>%
  dplyr::rename(sample=tube_number) %>%
  separate(region, into=c('chr', 'coord'), sep=':') %>%
  separate(coord, into = c('start', 'end'), sep='-')

region_stats = function(vcf_df){
  sample_i = vcf_df %>% pull(sample) %>% unique()
  
  region = region_df %>%
    filter(sample == sample_i)
  
  region
  
  if (nrow(region) == 0){
    return(NULL)
  } else{
  summarize_genomic_origin(vcf_df,
                           region_chr=region$chr,
                           region_start=region$start,
                           region_end=region$end)
  }
}

region_wgs_var_summaries = map(group_split(input_vcf_tables_df_list), region_stats)

```

