---
title: "create_pipeline_samplesheet"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{create_pipeline_samplesheet}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r setup,message=FALSE, warning = FALSE}
library(BSA)
library(tidyverse)
```

# Parse data from Daniel's full metadata sheet
This is a subset of only the BSA6 samples. Goal is to parse the data into 
a the appropriate format for the NF pipeline. Required columns are: 

- sample (a unique name)
- group (this will be used to process bam files together in variant calling step. 
    group 0 signifies that the file will only be processed individually. Any other 
    number may be used to identify other groups, eg if there are 5 files, one 
    file to be processed individually, and then groups of 2, group might look 
    like 0,1,1,2,2. Note that all bam files are processed individually by default)
- pool (pool from which the sample originates. not unique across pools, eg P1-P5 and Pall)
- replicate (replicate of a given pool, eg 1,2,3)
- day (number of days passed in time course before harvest)
- cond (culture condition, eg inoculum, lung, brain, ypd)
- experiment (unique identifier of the experiment, eg BSA6)
- runNumber (from the sequencer, eg 4726)
- fastq_1 and fastq_2 paths to the fastq files

```{r, message=FALSE, warning = FALSE}
relevel_cond = function(cond){
  switch (cond,
  L = "lung",
  Y = "ypd",
  B = "brain",
  I = "inoculum"
)
}


df = readRDS(system.file("bsa6_raw_meta.rds", package = "BSA")) %>%
  mutate(pool = ifelse(str_detect(`Strain Alias`, "all"), 
                       "all", 
                       str_remove(str_extract(`Strain Alias`, "^P\\d"), "^P"))) %>%
  mutate(tmp = str_remove(`Strain Alias`, paste0("P", pool))) %>%
  mutate(cond = ifelse(substr(tmp,1,1) == "1", NA, substr(tmp,1,1))) %>%
  mutate(day = ifelse(is.na(cond), 15, 8)) %>%
  mutate(tmp = str_remove(tmp, as.character(day))) %>%
  mutate(cond = ifelse(is.na(cond), substr(tmp,1,1), cond)) %>%
  mutate(day = ifelse(day == 8 & cond == "Y", 1, day)) %>%
  mutate(day = ifelse(day == 8 & cond == "I", 0, day)) %>%
  mutate(tmp = str_remove(tmp,cond)) %>%
  dplyr::rename(replicate = tmp) %>%
  mutate(replicate = ifelse(replicate == "", 1,replicate)) %>%
  dplyr::rename(runNumber = `Run number`) %>%
  mutate(Experiment = str_remove_all(str_replace(Experiment, "-", "_"), " ")) %>%
  dplyr::rename(experiment = Experiment) %>%
  mutate(fastq_1 = file.path(FileFolder,FirstPairFileName)) %>%
  dplyr::rename(sample = `Strain Alias`) %>%
  mutate(group = ifelse(str_detect(sample, "all"), "oneMouse", "sepMouse")) %>%
  mutate(fastq_2 = str_replace(fastq_1, "_R1_", "_R2_")) %>%
  mutate(cond = unlist(map(cond, relevel_cond))) %>%
  dplyr::select(sample,group, pool, cond, day, replicate, experiment, 
                runNumber, fastq_1, fastq_2) %>%
  mutate(fastq_1 = file.path("data",basename(fastq_1)),
         fastq_2 = file.path("data",basename(fastq_2)))

write_csv(df, "/mnt/scratch/variant_calling_pipeline/bsa6_samplesheet.csv")
# write_tsv(df, "/mnt/scratch/variant_calling_pipeline/bsa6_samplesheet.tsv")
```

```{r}
df
```

