<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A wrapper for all the functions that perform the full G prime analysis to
identify QTL. The following steps are performed: 1) Genome-wide G
statistics are calculated by getG.  G is defined by the
equation: $$G = 2*\sum_{i=1}^{4} n_{i}*ln\frac{obs(n_i)}{exp(n_i)}$$ Where for each SNP, \(n_i\) from i = 1
to 4 corresponds to the reference and alternate allele depths for each bulk,
as described in the following table:
AlleleHigh BulkLow Bulk
Reference\(n_1\)\(n_2\)
Alternate\(n_3\)\(n_4\)

...and \(obs(n_i)\) are the observed allele
depths as described in the data frame. getG calculates the G statistic
using expected values assuming read depth is equal for all alleles in both
bulks: $$exp(n_1) = ((n_1 + n_2)*(n_1 + n_3))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_2) = ((n_2 + n_1)*(n_2 + n_4))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_3) = ((n_3 + n_1)*(n_3 + n_4))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_4) = ((n_4 + n_2)*(n_4 + n_3))/(n_1 + n_2 + n_3 + n_4)$$ 2) G'
A tricube-smoothed G statistic is predicted by local regression within each
chromosome using tricubeStat. This works as a weighted average
across neighboring SNPs that accounts for Linkage disequilibrium (LD) while
minizing noise attributed to SNP calling errors. G values for neighboring
SNPs within the window are weighted by physical distance from the focal SNP.
  3) P-values are estimated based using the non-parametric method
described by Magwene et al. 2011 with the function getPvals.
Breifly, using the natural log of Gprime a median absolute deviation (MAD) is
calculated. The Gprime set is trimmed to exclude outlier regions (i.e. QTL)
based on Hampel's rule. An alternate method for filtering out QTL is proposed
using absolute delta SNP indeces greater than 0.1 to filter out potential
QTL. An estimation of the mode of the trimmed set is calculated using the
mlv function from the package modeest. Finally, the
mean and variance of the set are estimated using the median and mode and
p-values are estimated from a log normal distribution.   4) Negative
Log10- and Benjamini-Hochberg adjusted p-values are calculated using
p.adjust

"><title>Identify QTL using a smoothed G statistic — runQTLseqAnalysis_local • BSA</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.8/font.css" rel="stylesheet"><link href="../deps/JetBrains_Mono-0.4.8/font.css" rel="stylesheet"><link href="../deps/Roboto_Slab-0.4.8/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identify QTL using a smoothed G statistic — runQTLseqAnalysis_local"><meta property="og:description" content="A wrapper for all the functions that perform the full G prime analysis to
identify QTL. The following steps are performed: 1) Genome-wide G
statistics are calculated by getG.  G is defined by the
equation: $$G = 2*\sum_{i=1}^{4} n_{i}*ln\frac{obs(n_i)}{exp(n_i)}$$ Where for each SNP, \(n_i\) from i = 1
to 4 corresponds to the reference and alternate allele depths for each bulk,
as described in the following table:
AlleleHigh BulkLow Bulk
Reference\(n_1\)\(n_2\)
Alternate\(n_3\)\(n_4\)

...and \(obs(n_i)\) are the observed allele
depths as described in the data frame. getG calculates the G statistic
using expected values assuming read depth is equal for all alleles in both
bulks: $$exp(n_1) = ((n_1 + n_2)*(n_1 + n_3))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_2) = ((n_2 + n_1)*(n_2 + n_4))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_3) = ((n_3 + n_1)*(n_3 + n_4))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_4) = ((n_4 + n_2)*(n_4 + n_3))/(n_1 + n_2 + n_3 + n_4)$$ 2) G'
A tricube-smoothed G statistic is predicted by local regression within each
chromosome using tricubeStat. This works as a weighted average
across neighboring SNPs that accounts for Linkage disequilibrium (LD) while
minizing noise attributed to SNP calling errors. G values for neighboring
SNPs within the window are weighted by physical distance from the focal SNP.
  3) P-values are estimated based using the non-parametric method
described by Magwene et al. 2011 with the function getPvals.
Breifly, using the natural log of Gprime a median absolute deviation (MAD) is
calculated. The Gprime set is trimmed to exclude outlier regions (i.e. QTL)
based on Hampel's rule. An alternate method for filtering out QTL is proposed
using absolute delta SNP indeces greater than 0.1 to filter out potential
QTL. An estimation of the mode of the trimmed set is calculated using the
mlv function from the package modeest. Finally, the
mean and variance of the set are estimated using the median and mode and
p-values are estimated from a log normal distribution.   4) Negative
Log10- and Benjamini-Hochberg adjusted p-values are calculated using
p.adjust

"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BSA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BSA3.html">BSA3</a>
    <a class="dropdown-item" href="../articles/BSA6.html">BSA6</a>
    <a class="dropdown-item" href="../articles/BSA_run_21.html">BSA_run_21</a>
    <a class="dropdown-item" href="../articles/CreatePipelineSamplesheet.html">CreatePipelineSamplesheet</a>
    <a class="dropdown-item" href="../articles/bsa_strain_genotype_check.html">bsa_strain_genotype_check</a>
    <a class="dropdown-item" href="../articles/c8_incuding_conditions_rnaseq.html">c8_incuding_conditions_rnaseq</a>
    <a class="dropdown-item" href="../articles/c8_variants.html">c8_variants</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmatKhan/BSA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Identify QTL using a smoothed G statistic</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmatKhan/BSA/blob/HEAD/R/runQTLseqAnalysis.R" class="external-link"><code>R/runQTLseqAnalysis.R</code></a></small>
      <div class="d-none name"><code>runQTLseqAnalysis_local.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A wrapper for all the functions that perform the full G prime analysis to
identify QTL. The following steps are performed:<br> 1) Genome-wide G
statistics are calculated by <code>getG</code>. <br> G is defined by the
equation: $$G = 2*\sum_{i=1}^{4} n_{i}*ln\frac{obs(n_i)}{exp(n_i)}$$ Where for each SNP, \(n_i\) from i = 1
to 4 corresponds to the reference and alternate allele depths for each bulk,
as described in the following table:</p><table class="table table"><tr><td>Allele</td><td>High Bulk</td><td>Low Bulk</td></tr><tr><td>Reference</td><td>\(n_1\)</td><td>\(n_2\)</td></tr><tr><td>Alternate</td><td>\(n_3\)</td><td>\(n_4\)</td></tr></table><p>...and \(obs(n_i)\) are the observed allele
depths as described in the data frame. <code>getG</code> calculates the G statistic
using expected values assuming read depth is equal for all alleles in both
bulks: $$exp(n_1) = ((n_1 + n_2)*(n_1 + n_3))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_2) = ((n_2 + n_1)*(n_2 + n_4))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_3) = ((n_3 + n_1)*(n_3 + n_4))/(n_1 + n_2 + n_3 + n_4)$$
$$exp(n_4) = ((n_4 + n_2)*(n_4 + n_3))/(n_1 + n_2 + n_3 + n_4)$$<br> 2) G'</p><ul><li><p>A tricube-smoothed G statistic is predicted by local regression within each
chromosome using <code>tricubeStat</code>. This works as a weighted average
across neighboring SNPs that accounts for Linkage disequilibrium (LD) while
minizing noise attributed to SNP calling errors. G values for neighboring
SNPs within the window are weighted by physical distance from the focal SNP.
<br><br> 3) P-values are estimated based using the non-parametric method
described by Magwene et al. 2011 with the function <code>getPvals</code>.
Breifly, using the natural log of Gprime a median absolute deviation (MAD) is
calculated. The Gprime set is trimmed to exclude outlier regions (i.e. QTL)
based on Hampel's rule. An alternate method for filtering out QTL is proposed
using absolute delta SNP indeces greater than 0.1 to filter out potential
QTL. An estimation of the mode of the trimmed set is calculated using the
<code><a href="https://rdrr.io/pkg/modeest/man/mlv.html" class="external-link">mlv</a></code> function from the package modeest. Finally, the
mean and variance of the set are estimated using the median and mode and
p-values are estimated from a log normal distribution. <br><br> 4) Negative
Log10- and Benjamini-Hochberg adjusted p-values are calculated using
<code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code></p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runQTLseqAnalysis_local</span><span class="op">(</span></span>
<span>  <span class="va">SNPset</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  popStruc <span class="op">=</span> <span class="st">"F2"</span>,</span>
<span>  <span class="va">bulkSize</span>,</span>
<span>  depth <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  replications <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  filter <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  intervals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">95</span>, <span class="fl">99</span><span class="op">)</span>,</span>
<span>  chrom_col <span class="op">=</span> <span class="st">"CHROM"</span>,</span>
<span>  coordinate_col <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>  delta_snp_col <span class="op">=</span> <span class="st">"deltaSNP"</span>,</span>
<span>  dp_low_col <span class="op">=</span> <span class="st">"DP.LOW"</span>,</span>
<span>  dp_high_col <span class="op">=</span> <span class="st">"DP.HIGH"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>SNPset</dt>
<dd><p>Data frame SNP set containing previously filtered SNPs</p></dd>


<dt>windowSize</dt>
<dd><p>the window size (in base pairs) bracketing each SNP for which
to calculate the statitics. Magwene et. al recommend a window size of ~25
cM, but also recommend optionally trying several window sizes to test if
peaks are over- or undersmoothed.</p></dd>


<dt>popStruc</dt>
<dd><p>the population structure. Defaults to "F2" and assumes "RIL" otherwise.</p></dd>


<dt>bulkSize</dt>
<dd><p>non-negative integer vector. The number of individuals in
each simulated bulk. Can be of length 1, then both bulks are set to the
same size. Assumes the first value in the vector is the simulated high
bulk.</p></dd>


<dt>depth</dt>
<dd><p>This is calculated in the script, usually. It is input to a
QTLSeqR function simulateConfInt_local</p></dd>


<dt>replications</dt>
<dd><p>integer. The number of bootstrap replications.</p></dd>


<dt>filter</dt>
<dd><p>numeric. A minimum SNP-index filter</p></dd>


<dt>intervals</dt>
<dd><p>confidence intervals -- note this is part of daniel's
additional code to the QTLseqR package</p></dd>


<dt>chrom_col</dt>
<dd><p>name of the column which stores the chromosome label.
Default is CHROM</p></dd>


<dt>coordinate_col</dt>
<dd><p>name of the column which stores the snp coordinate.
Default is POS</p></dd>


<dt>delta_snp_col</dt>
<dd><p>name of the column which stores the change in allele
frequency. Default is deltaSNP</p></dd>


<dt>dp_low_col</dt>
<dd><p>name of the column which stores the dp_low value. Default
DP.LOW</p></dd>


<dt>dp_high_col</dt>
<dd><p>name of the column which stores the dp_high value.
Default is DP.HIGH</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The supplied SNP set tibble after G' analysis. Includes five new
columns:</p><ul><li><p>Gprime -
The tricube smoothed G statistic based on the supplied window size</p></li>
<li><p>pvalue - the pvalue at each SNP calculatd by non-parametric
estimation</p></li>
<li><p>negLog10Pval - the -Log10(pvalue) supplied for quick
plotting</p></li>
<li><p>qvalue - the Benajamini-Hochberg adjusted p-value</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>DETAILS</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>THIS IS NEARLY A VERBATIM COPY OF QTLseqR</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Daniel Agustihno, Chase Mateusiak.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

