<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BSA">
<title>c8_variants â€¢ BSA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.7/font.css" rel="stylesheet">
<link href="../deps/JetBrains_Mono-0.4.7/font.css" rel="stylesheet">
<link href="../deps/Roboto_Slab-0.4.7/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="c8_variants">
<meta property="og:description" content="BSA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BSA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BSA3.html">BSA3</a>
    <a class="dropdown-item" href="../articles/BSA6.html">BSA6</a>
    <a class="dropdown-item" href="../articles/CreatePipelineSamplesheet.html">CreatePipelineSamplesheet</a>
    <a class="dropdown-item" href="../articles/c8_incuding_conditions_rnaseq.html">c8_incuding_conditions_rnaseq</a>
    <a class="dropdown-item" href="../articles/c8_variants.html">c8_variants</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmatKhan/BSA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>c8_variants</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmatKhan/BSA/blob/HEAD/vignettes/c8_variants.Rmd" class="external-link"><code>vignettes/c8_variants.Rmd</code></a></small>
      <div class="d-none name"><code>c8_variants.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cmatKhan/BSA" class="external-link">BSA</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Give your self a nice description of what your script does -- this will be</span></span>
<span><span class="co"># helpful to you-in-the-future</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># This script serves as an example of how to extract data from a VCF file into</span></span>
<span><span class="co"># a more usable format. In this case, it turns a VCF file into a GDS file,</span></span>
<span><span class="co"># and then uses SeqArray and SeqVarTools to extract the variant info (chr, pos,</span></span>
<span><span class="co"># depth, annotations, ...) and parse that data into a dataframe. The final</span></span>
<span><span class="co"># product of this script is a dataframe that can be used to filter variants</span></span>
<span><span class="co"># in the DK samples based on how their genotypes compare to the YSP2-8 genotype</span></span>
<span><span class="co"># at the same location.</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zhengxwen/SeqArray" class="external-link">SeqArray</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/smgogarten/SeqVarTools" class="external-link">SeqVarTools</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#' @title Tidy data from gds file</span></span>
<span><span class="co">#' @description Tidy data to long format with column sample, variant, and the</span></span>
<span><span class="co">#'   metric name. internal use only</span></span>
<span><span class="co">#'</span></span>
<span><span class="co">#' @param data_mat a data matrix in sample x variant format</span></span>
<span><span class="co">#' @param values_cname name of the metric -- eg RealDepth</span></span>
<span><span class="co">#' @param rnames vector to name the columns, eg sample.ids</span></span>
<span><span class="co">#' @param cnames vector to name the rows, eg variant_ids</span></span>
<span><span class="co">#'</span></span>
<span><span class="co">#' @return a tidy dataframe in long format with columns sample, variant, metric</span></span>
<span><span class="co">#'</span></span>
<span><span class="co">#' @importFrom dplyr as_tibble</span></span>
<span><span class="co">#' @importFrom tidyr pivot_longer</span></span>
<span><span class="va">tidy_metrics</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">data_mat</span>, <span class="va">values_cname</span>, <span class="va">rnames</span>, <span class="va">cnames</span><span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>  <span class="va">errors</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    dimension_error <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"vcf_to_qtlseqr_table error: rownames do not "</span>,</span>
<span>                             <span class="st">"match data matrix. see tidy_metrics() function"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">rnames</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">data_mat</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cnames</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">data_mat</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="va">errors</span><span class="op">$</span><span class="va">dimension_error</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="va">data_mat</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">`rownames&lt;-`</span><span class="op">(</span><span class="va">rnames</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">`colnames&lt;-`</span><span class="op">(</span><span class="va">cnames</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="va">sample</span>, names_to <span class="op">=</span> <span class="st">"variant"</span>, values_to <span class="op">=</span> <span class="va">values_cname</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># this is already done -- just load the allClinical.ann.gds</span></span>
<span><span class="co"># tryCatch(</span></span>
<span><span class="co">#   SeqArray::seqVCF2GDS(</span></span>
<span><span class="co">#     here("allClinical.ann.vcf"),</span></span>
<span><span class="co">#     here("inst/allClinical.ann.gds")),</span></span>
<span><span class="co">#   error = function(e){</span></span>
<span><span class="co">#     message(e)</span></span>
<span><span class="co">#   })</span></span>
<span></span>
<span></span>
<span><span class="co"># After creating the GDS file, we now 'open' it. Note that this does not</span></span>
<span><span class="co"># open the entire file, but rather creates a file handle. The difference is</span></span>
<span><span class="co"># that we are not reading the data into RAM. This means that we can handle</span></span>
<span><span class="co"># enormous VCF files, which is a common use case with this type of data</span></span>
<span><span class="va">all_strains_gds</span> <span class="op">=</span> <span class="fu">SeqArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeqArray/man/seqOpen.html" class="external-link">seqOpen</a></span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">"inst/allClinical.ann.gds"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set a filter such that we only look at loci with at most 1 alternative allele</span></span>
<span><span class="co"># In the all_strains data, there is only a single position with more than 1 alt</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Setting single allele loci filter..."</span><span class="op">)</span></span>
<span><span class="va">single_allele_fltr</span> <span class="op">=</span></span>
<span>  <span class="fu">SeqArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeqArray/man/seqGetData.html" class="external-link">seqGetData</a></span><span class="op">(</span><span class="va">all_strains_gds</span>,</span>
<span>                       <span class="st">'variant.id'</span><span class="op">)</span><span class="op">[</span><span class="fu">SeqVarTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeqVarTools/man/allele-methods.html" class="external-link">nAlleles</a></span><span class="op">(</span><span class="va">all_strains_gds</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"...total variants: %s"</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu">SeqVarTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeqVarTools/man/allele-methods.html" class="external-link">nAlleles</a></span><span class="op">(</span><span class="va">all_strains_gds</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"...FALSE refers to the number of multi allelic loci"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">SeqVarTools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeqVarTools/man/allele-methods.html" class="external-link">nAlleles</a></span><span class="op">(</span><span class="va">all_strains_gds</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">SeqArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeqArray/man/seqSetFilter.html" class="external-link">seqSetFilter</a></span><span class="op">(</span><span class="va">all_strains_gds</span>,</span>
<span>                       variant.id<span class="op">=</span><span class="va">single_allele_fltr</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select only KN99a and TDY1993 (C8)</span></span>
<span><span class="co"># SeqArray::seqSetFilter(</span></span>
<span><span class="co">#   all_strains_gds, </span></span>
<span><span class="co">#   sample.sel = SeqArray::seqGetData(</span></span>
<span><span class="co">#     all_strains_gds, 'sample.id') %in% c('KN99a', 'TDY1993'))</span></span>
<span></span>
<span><span class="co"># This is the set of 'fields' we're going to extract from the GDS file. Note</span></span>
<span><span class="co"># that the return type of each of these 'fields' will not be the same -- eg,</span></span>
<span><span class="co"># genotype returns a 3 dimensional matrix, chromosome returns a vector, ...</span></span>
<span><span class="va">data_to_extract</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chromosome"</span>, <span class="st">"position"</span>, <span class="st">"$ref"</span>, <span class="st">"$alt"</span>,</span>
<span>                    <span class="st">"annotation/qual"</span>, <span class="st">"genotype"</span>, <span class="st">"annotation/info/ANN"</span>,</span>
<span>                    <span class="st">"annotation/info/DP"</span>, <span class="st">"annotation/format/DP"</span>,</span>
<span>                    <span class="st">"annotation/format/RO"</span>, <span class="st">"annotation/format/AO"</span>,</span>
<span>                    <span class="st">"sample.id"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># `seqGetData` is one of the methods we can use to extract data from the</span></span>
<span><span class="co"># GDS file. Using `seqGetData` extracts the data from the file and returns it</span></span>
<span><span class="co"># as a list in memory. In the example below, we now have the annotations data</span></span>
<span><span class="co"># in memory and can start to parse it</span></span>
<span><span class="va">all_strains_annotations</span> <span class="op">=</span> <span class="fu">SeqArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SeqArray/man/seqGetData.html" class="external-link">seqGetData</a></span><span class="op">(</span></span>
<span>  <span class="va">all_strains_gds</span>, <span class="va">data_to_extract</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create unique IDs for the variants</span></span>
<span><span class="va">variant_ids</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var_"</span>,</span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`annotation/format/DP`</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># The annotations are returned in a list of two items, `length` and `data`.</span></span>
<span><span class="co"># The `length` list looks something like 3, 1, 6, 2, ... where the number</span></span>
<span><span class="co"># represents the number of times the variant at that index is annotated to a</span></span>
<span><span class="co"># different spnEFF annotation. These annotations are stored in the list `data`.</span></span>
<span><span class="co"># So, variant 1 has 3 different annotations in `data`, variant 2 has 1</span></span>
<span><span class="co"># annotation, and so on.</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># The goal of the code below is to extract a "key" column from the</span></span>
<span><span class="co"># annotations to the variants. If we consider the variant index to be the key,</span></span>
<span><span class="co"># what we want to know is that the first 3 entries of the `data` map to the</span></span>
<span><span class="co"># variant at index 1</span></span>
<span><span class="va">variant_index</span> <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html" class="external-link">map2</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`annotation/info/ANN`</span><span class="op">$</span><span class="va">length</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`annotation/info/ANN`</span><span class="op">$</span><span class="va">length</span>,</span>
<span>  <span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">.x</span>,<span class="va">.y</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a dataframe where the first column is the variant name, the second</span></span>
<span><span class="co"># column is the full annotation string, and the `alt`, `effect`, `impact`,</span></span>
<span><span class="co"># and `gene` columns are created by parsing out the relevant information from</span></span>
<span><span class="co"># the annotation string</span></span>
<span></span>
<span><span class="va">variant_annotation_df</span> <span class="op">=</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  variant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">'var'</span>, <span class="va">variant_index</span>, sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span>,</span>
<span>  annotation <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`annotation/info/ANN`</span><span class="op">$</span><span class="va">data</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">annotation</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'alt'</span>, <span class="st">'effect'</span>, <span class="st">'impact'</span>, <span class="st">'col4'</span>, <span class="st">'gene'</span>, </span>
<span>                                <span class="st">'col6'</span>, <span class="st">'col7'</span>, <span class="st">'protein_coding_effect'</span>, </span>
<span>                                <span class="st">'col9'</span>, <span class="st">'col10'</span>, <span class="st">'aa'</span>, <span class="st">'col12'</span>, <span class="st">'col13'</span>, </span>
<span>                                <span class="st">'col14'</span>, <span class="st">'col15'</span><span class="op">)</span>, </span>
<span>             sep <span class="op">=</span> <span class="st">"\\|"</span>, fill <span class="op">=</span> <span class="st">"right"</span>,remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">variant</span>, <span class="va">annotation</span>, <span class="va">alt</span>, <span class="va">effect</span>, <span class="va">impact</span>, <span class="va">gene</span>, </span>
<span>         <span class="va">protein_coding_effect</span>, <span class="va">aa</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>protein_coding_effect <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>    <span class="va">protein_coding_effect</span> <span class="op">==</span> <span class="st">'protein_coding'</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># this is part of a function from the BSA package which was set up as a</span></span>
<span><span class="co"># convenience to parse some of this data. It is reused here -- see</span></span>
<span><span class="co"># R/tidy_metrics.R</span></span>
<span><span class="va">data_to_transform</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  genotype     <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">genotype</span><span class="op">[</span><span class="fl">1</span>,,<span class="op">]</span>,</span>
<span>  RealDepth    <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`annotation/format/DP`</span>,</span>
<span>  Reference    <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`annotation/format/RO`</span>,</span>
<span>  Alternative1 <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`annotation/format/AO`</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co"># result is a long data frame with columns sample, variant, RealDepth,</span></span>
<span><span class="co"># Reference (which is the reference allele depth) and Alternative1, which is</span></span>
<span><span class="co"># the Alternative depth. Note that if single_allele_loci_only is not set,</span></span>
<span><span class="co"># then this is not actually Alternative1, but just Alternative</span></span>
<span><span class="va">depth_metrics_df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html" class="external-link">map2</a></span><span class="op">(</span><span class="va">data_to_transform</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data_to_transform</span><span class="op">)</span>,</span>
<span>              <span class="va">tidy_metrics</span>,</span>
<span>              <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">sample.id</span>,</span>
<span>              <span class="va">variant_ids</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/reduce.html" class="external-link">reduce</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="va"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now we're going to create our analysis data -- in this case, we're going to</span></span>
<span><span class="co"># join the depth df to a temporary dataframe with the chromosome, position, etc</span></span>
<span><span class="co"># information. Then we join the annotations dataframe and filter for positions</span></span>
<span><span class="co"># which have RealDepth (actual reads aligning over a given position,</span></span>
<span><span class="co"># not counting reads which aligned with a gap over that variant position) of</span></span>
<span><span class="co"># at least 10</span></span>
<span><span class="va">all_strains_annotated_df</span> <span class="op">=</span> <span class="va">depth_metrics_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>variant <span class="op">=</span> <span class="va">variant_ids</span>,</span>
<span>         chr <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">chromosome</span>,</span>
<span>         pos <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">position</span>,</span>
<span>         ref <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`$ref`</span>,</span>
<span>         alt <span class="op">=</span> <span class="va">all_strains_annotations</span><span class="op">$</span><span class="va">`$alt`</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">variant_annotation_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>              <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">annotation</span>, <span class="op">-</span><span class="va">alt</span><span class="op">)</span>,</span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'variant'</span><span class="op">)</span>,</span>
<span>            relationship <span class="op">=</span> <span class="st">'many-to-many'</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">RealDepth</span><span class="op">)</span> <span class="op">&amp;</span> <span class="va">RealDepth</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># exclude any marker sequences</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">chr</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NAT"</span>, <span class="st">"G418"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># factor the `impact` column -- note that I ran everything up to the</span></span>
<span>  <span class="co"># `filter` step above, and then looked at the unique values in `impact`</span></span>
<span>  <span class="co"># to figure out the levels. The order of the labels determines how</span></span>
<span>  <span class="co"># this column will sort, with "HIGH" being highest impact and "MODIFIER"</span></span>
<span>  <span class="co"># the lowest</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>impact <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">impact</span>,levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'HIGH'</span>, <span class="st">'MODERATE'</span>,</span>
<span>                                           <span class="st">'LOW'</span>, <span class="st">'MODIFIER'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># `HIGH` on top</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">impact</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now we're going to extract just the tdy1993 data and reform it into a</span></span>
<span><span class="co"># dataframe with columns `variant` and `tdy1993_genotype`</span></span>
<span><span class="va">tdy1993_genotype_df</span> <span class="op">=</span> <span class="va">all_strains_annotated_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample</span> <span class="op">==</span> <span class="st">'TDY1993'</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">variant</span>, <span class="va">genotype</span>, <span class="va">RealDepth</span>, <span class="va">Reference</span>, <span class="va">Alternative1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>tdy1993_genotype <span class="op">=</span> <span class="va">genotype</span>,</span>
<span>                tdy1993_RealDepth <span class="op">=</span> <span class="va">RealDepth</span>,</span>
<span>                tdy1993_Reference <span class="op">=</span> <span class="va">Reference</span>,</span>
<span>                tdy1993_Alternative1 <span class="op">=</span> <span class="va">Alternative1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read in gene info (parsed from the gtf from fungidb)</span></span>
<span><span class="va">kn99_genes</span> <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">"inst/kn99_genes.csv"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">chr</span>, <span class="va">source</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>gene_start <span class="op">=</span> <span class="va">start</span>,</span>
<span>                gene_end <span class="op">=</span> <span class="va">end</span>,</span>
<span>                gene_strand <span class="op">=</span> <span class="va">strand</span>,</span>
<span>                gene_description <span class="op">=</span> <span class="va">description</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># And finally, we remove the tdy1993 data from the all_strains data, and then join</span></span>
<span><span class="co"># the ysp2_genotype_df so that we now have an additional column called</span></span>
<span><span class="co"># `ysp_2_8_genotype`</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># We filter for "complete cases" (or, only keep rows with no NAs in any column)</span></span>
<span><span class="co"># so that we're only looking at variants for which we have reliable data over</span></span>
<span><span class="co"># both a given sample, and tdy1993</span></span>
<span><span class="va">analysis_df</span> <span class="op">=</span> <span class="va">all_strains_annotated_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sample</span> <span class="op">==</span> <span class="st">'KN99a'</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">tdy1993_genotype_df</span>,</span>
<span>            by <span class="op">=</span> <span class="st">'variant'</span>,</span>
<span>            relationship <span class="op">=</span> <span class="st">'many-to-many'</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html" class="external-link">complete.cases</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genotype</span> <span class="op">!=</span> <span class="fl">0</span> <span class="op">|</span> <span class="va">tdy1993_genotype</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>kn99a_genotype <span class="op">=</span> <span class="va">genotype</span>,</span>
<span>                kn99a_RealDepth <span class="op">=</span> <span class="va">RealDepth</span>,</span>
<span>                kn99a_Reference <span class="op">=</span> <span class="va">Reference</span>,</span>
<span>                kn99a_Alternative1 <span class="op">=</span> <span class="va">Alternative1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">kn99_genes</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'gene'</span> <span class="op">=</span> <span class="st">'ID'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">variant</span>, <span class="va">chr</span>, <span class="va">pos</span>, <span class="va">kn99a_genotype</span>, </span>
<span>                <span class="va">tdy1993_genotype</span>,<span class="va">ref</span>,<span class="va">alt</span>,<span class="va">impact</span>,<span class="va">effect</span>,<span class="va">aa</span>,<span class="va">gene</span>, </span>
<span>                <span class="va">protein_coding_effect</span>, <span class="va">gene_start</span>, <span class="va">gene_end</span>, <span class="va">gene_strand</span>, </span>
<span>                <span class="va">gene_description</span>, <span class="va">kn99a_RealDepth</span>,<span class="va">kn99a_Reference</span>, </span>
<span>                <span class="va">kn99a_Alternative1</span>, <span class="va">tdy1993_RealDepth</span>, <span class="va">tdy1993_Reference</span>, </span>
<span>                <span class="va">tdy1993_Alternative1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nstrains_per_variant</span> <span class="op">=</span> <span class="va">all_strains_annotated_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">sample</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'KN99a'</span>, <span class="st">'TDY1993'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">variant</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">tally</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># write results</span></span>
<span><span class="va">analysis_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">impact</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>tally<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>            effect_all<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">effect</span><span class="op">)</span>,collapse<span class="op">=</span><span class="st">","</span><span class="op">)</span>, </span>
<span>            aa_all<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">aa</span><span class="op">)</span>,collapse <span class="op">=</span> <span class="st">','</span><span class="op">)</span>,</span>
<span>            gene_description <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gene_description</span><span class="op">)</span>,</span>
<span>            gene_start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gene_start</span><span class="op">)</span>,</span>
<span>            gene_end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gene_end</span><span class="op">)</span>,</span>
<span>            gene_strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gene_strand</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene</span>,<span class="va">gene_description</span>,<span class="va">gene_start</span>,<span class="va">gene_end</span>,<span class="va">gene_strand</span>,</span>
<span>         <span class="va">impact</span>,<span class="va">tally</span>,<span class="va">effect_all</span>,<span class="va">aa_all</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="fu">here</span><span class="op">(</span><span class="st">'results/gene_overview_kn99a_vs_c8_20230726.csv'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="va">analysis_df</span>, <span class="fu">here</span><span class="op">(</span><span class="st">'results/kn99a_vs_c8_20230725.csv'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Daniel Agustihno, Chase Mateusiak.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
