<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BSA">
<title>processing_instructions • BSA</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="processing_instructions">
<meta property="og:description" content="BSA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BSA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BSA2.html">BSA2</a>
    <a class="dropdown-item" href="../articles/BSA2_plots.html">BSA2_plots</a>
    <a class="dropdown-item" href="../articles/processing_instructions.html">processing_instructions</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmatKhan/BSA/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>processing_instructions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmatKhan/BSA/blob/HEAD/vignettes/processing_instructions.Rmd" class="external-link"><code>vignettes/processing_instructions.Rmd</code></a></small>
      <div class="d-none name"><code>processing_instructions.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h2>
<p>The first step is to align the reads. I normally name the samples according with what they are, so that’s are represented in the fastq file. The first step for me is to createan ID file containing name of the sample and file full address, separated by space.You can change that to other sep in the script. I was considering changing the sep toa comma, to be easily exported from a Excel. But haven’t done that. Having the name of the sample in the fastq file allows me to use a simple script tomake this ID file. For BSA2 samples, you have to look at the excel sheet DanielSeqDatabase.xlsx to find where the files of interest are. Then we would just do this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> /lts/mblab/personal/daniel.agustinho/raw/BSA2/<span class="pp">*</span>.fastq.gz/scratch/MyFolder</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /scratch/MyFolder/scratch/mblab/daniel.agustinho/tools/IDmaker.sh ID</span></code></pre></div>
<div class="section level3">
<h3 id="idmaker-sh">IDmaker.sh<a class="anchor" aria-label="anchor" href="#idmaker-sh"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[chasem@login</span> ~]$ cat /scratch/mblab/daniel.agustinho/tools/IDmaker.sh</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="va">output</span><span class="op">=</span><span class="va">$1</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="pp">*</span>R1_001.fastq.gz <span class="kw">|</span><span class="fu">awk</span> <span class="at">-F</span> <span class="st">'_'</span> <span class="st">'{print $3}'</span> <span class="op">&gt;</span> temp1.txt</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-d</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">"</span>/<span class="pp">*</span>R1_001.fastq.gz <span class="op">&gt;</span> temp2.txt</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="ex">paste</span> <span class="at">-d</span> <span class="st">" "</span> temp1.txt temp2.txt <span class="op">&gt;</span> <span class="va">$output</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> temp1.txt temp2.txt</span></code></pre></div>
<p>This will create a file called ID containing the name of the sample. Copy that and the fastq.gz files to your analysis folder in scratch. Create some useful folders that will beused by the scripts. Call alignments with NGM (you can use another). Observe thatthe script has already the “KN99 reference” that’s older. Feel free to adjust that inyour own copied version of the script.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> bams freebayes/indvVCF </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> raw log SVssbatch <span class="at">--array</span><span class="op">=</span>1-<span class="va">$(</span><span class="fu">wc</span> <span class="at">-l</span> <span class="op">&lt;</span> ID<span class="va">)</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="ex">/scratch/mblab/daniel.agustinho/tools/ngm.job</span> ID bams KN99</span></code></pre></div>
<p>This will create alignment files (bam and bai) for each sample in the bams folder. Itwill also call CNVnator for each one of them, generating tables and CNV VCFs with the same sample names in SVs/indvCNVs folder. It will call Delly for SVs, in theSVs/delly folder. But recently I observed that Delly sucks. I recommend using Manta instead. It will also call individual VCFs using freebayes and annotating the withSNPEff in the freebayes/indvVCF folder. If you have any problems or errors, check the log/ folder</p>
</div>
</div>
<div class="section level2">
<h2 id="variant-calling-for-bsa-analysis">Variant calling for BSA analysis<a class="anchor" aria-label="anchor" href="#variant-calling-for-bsa-analysis"></a>
</h2>
<p>When this is finished, you should call freebayes again for the population VCF. First, create a list of bams, and use those for the freebayes call</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-d</span> <span class="st">"</span><span class="va">$PWD</span><span class="st">"</span>/bams/<span class="pp">*</span>.merged.tagged.ngm.bam <span class="op">&gt;&gt;</span> bamList</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">/scratch/mblab/daniel.agustinho/tools/sbatch_maker/freebayes.bamList.sbatch</span> <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  KN99 <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  BSA2 <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  bamList</span></code></pre></div>
<div class="section level3">
<h3 id="freebayes-bamlist-sbatch">freebayes.bamList.sbatch<a class="anchor" aria-label="anchor" href="#freebayes-bamlist-sbatch"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[chasem@login</span> ~]$ cat /scratch/mblab/daniel.agustinho/tools/sbatch_maker/freebayes.bamList.sbatch</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -o freebayes-out</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -e freebayes-err</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=50000</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mail-type=</span><span class="re">END</span><span class="co">,FAIL</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --array 1-17%17</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="ex">ml</span> freebayes/1.1.0</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="ex">ml</span> vcftools/0.1.14</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="va">genome</span><span class="op">=</span><span class="va">$1</span> <span class="co"># Either H99 or KN99</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="va">name</span><span class="op">=</span><span class="va">$2</span> <span class="co">#This is the name of the clinical strain used for basis of the comparison</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="va">bamList</span><span class="op">=</span><span class="va">$3</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="bu">[</span> <span class="va">$genome</span> <span class="ot">=</span>  <span class="st">"H99"</span> <span class="bu">]</span><span class="kw">;</span><span class="cf">then</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a> <span class="va">reference</span><span class="op">=</span><span class="st">"/scratch/mblab/daniel.agustinho/references/crNeoH99.fasta"</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="cf">elif</span> <span class="bu">[</span> <span class="va">$genome</span> <span class="ot">=</span> <span class="st">"KN99"</span> <span class="bu">]</span><span class="kw">;</span><span class="cf">then</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a> <span class="va">reference</span><span class="op">=</span><span class="st">"/scratch/mblab/daniel.agustinho/references/crNeoKN99.fasta"</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Valid genomes are H99 or KN99. Review your 4th  argument."</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>    <span class="bu">exit</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="cf">fi</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="ex">/scratch/mblab/daniel.agustinho/tools/freebayes.bam_list.job</span> <span class="va">$reference</span> <span class="va">$name</span> <span class="va">$bamList</span> <span class="va">${SLURM_ARRAY_TASK_ID}</span></span></code></pre></div>
<p>This will create one VCF file per chromosome in the freebayes folder. Check if it ransmoothly (some samples may require more memmory), and if it did, we will merge allof those together using the VCFmerger.sh script. This script merges them in one bigVCF and deletes the VCF for individual chromosomes, so be sure the previous stepran smoothly. It also calls the next script, that makes tables from the VCF file that canbe easily read by the R script that will do the analysis.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scratch/mblab/daniel.agustinho/tools/VCFmerger.sh</span> BSA2</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="vcfmerger-sh">VCFmerger.sh<a class="anchor" aria-label="anchor" href="#vcfmerger-sh"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[chasem@login</span> ~]$ cat /scratch/mblab/daniel.agustinho/tools/VCFmerger.sh </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -o log/VCFmerger_%j-out</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -e log/VCFmerger_%j-err</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=50G</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="ex">ml</span> vcftools/0.1.14</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="ex">ml</span> R</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="va">name</span><span class="op">=</span><span class="va">$1</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> freebayes/VCFtables</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'#'</span> freebayes/freebayes.<span class="va">${name}</span>.chr1.vcf<span class="op">&gt;</span>freebayes/<span class="va">$name</span>.vcf</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> chr<span class="dt">{</span><span class="dv">1</span><span class="dt">..</span><span class="dv">14</span><span class="dt">}</span> chrM NAT G418</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span> <span class="fu">cat</span> freebayes/freebayes.<span class="va">${name}</span>.<span class="va">$i</span>.vcf <span class="kw">|</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="st">'^#'</span> <span class="op">&gt;&gt;</span> freebayes/<span class="va">$name</span>.vcf</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="ex">vcftools</span> <span class="at">--vcf</span> freebayes/<span class="va">${name}</span>.vcf <span class="at">--out</span> freebayes/<span class="va">${name}</span>.dusted.vcf <span class="at">--exclude-bed</span> /scratch/mblab/daniel.agustinho/references/dustmasked.KN99.bed <span class="at">--recode</span> <span class="at">--recode-INFO-all</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> freebayes/<span class="va">$name</span>.dusted.vcf.recode.vcf freebayes/<span class="va">$name</span>.clean.vcf</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> freebayes/freebayes.<span class="va">${name}</span>.chr<span class="pp">*</span>.vcf</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> freebayes/<span class="va">$name</span>.dusted.vcf.recode.vcf</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> freebayes/<span class="va">${name}</span>.vcf</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Starting the sorting.</span></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-e</span> <span class="st">"77s/#//"</span> freebayes/<span class="va">$name</span>.clean.vcf <span class="op">&gt;</span> freebayes/<span class="va">$name</span>.clean.vcf.test</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> /scratch/mblab/daniel.agustinho/tools/vcfSorter.R freebayes/<span class="va">$name</span>.clean.vcf.test</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'^##'</span> freebayes/<span class="va">$name</span>.clean.vcf  <span class="op">&gt;</span> freebayes/<span class="va">$name</span>.sorted.vcf <span class="co"># ${file/clean.vcf/sorted.vcf}</span></span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> data.vcf <span class="op">&gt;&gt;</span> freebayes/<span class="va">$name</span>.sorted.vcf</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="at">-e</span> <span class="st">"77s/CHROM/#CHROM/"</span> freebayes/<span class="va">$name</span>.sorted.vcf <span class="op">&gt;</span> freebayes/<span class="va">$name</span>.IGV.vcf</span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> data.vcf</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> freebayes/<span class="va">$name</span>.clean.vcf.test freebayes/<span class="va">$name</span>.clean.vcf freebayes/<span class="va">$name</span>.sorted.vcf</span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a><span class="ex">/scratch/mblab/daniel.agustinho/tools/VCF_tabler.sh</span> freebayes/<span class="va">$name</span>.IGV.vcf freebayes/VCFtables</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> freebayes/VCFtables/.txt</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a><span class="ex">/scratch/mblab/daniel.agustinho/tools/strainAnalyzer.sh</span></span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a><span class="ex">./popAnalyzer.sh</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="statistical-analysis">Statistical analysis<a class="anchor" aria-label="anchor" href="#statistical-analysis"></a>
</h2>
<p><strong>NOTE</strong> these scripts have been incorporated into the <code>R/</code> directory or other vignettes. Below is from Daniel:</p>
<p>That should create a BSA2.IGV.vcf file in the freebayes folder. You can look at it on IGV etc. It will also create a folder freebayes/VCFtables that contains one table (txtfile) for each sample that was called used as input for freebayes. Transfer those toyour personal laptop to analyse using the R script:</p>
<ul>
<li><code>BSA2Analysis.R</code></li>
</ul>
<p>Adjust the script to get the input from the proper folder (where the txt files you copied from the server are). This will generate a BSA2.RData file containing the analysis. You can use it in the second R script to make plots.</p>
<ul>
<li><code>BSA2plots.R</code></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by First Last.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
